README.md
# Flutter Text To Speech Improved

The `flutter_tts` plugin (https://github.com/dlutton/flutter_tts) is GREAT! Thanks [@dlutton](https://github.com/dlutton/) for all your hard work on getting it up and running. Now it is time to improve it by allowing for the Utterances API to report the 'progress' of the speech of the Utterance. This is relatively new for Android, only introduced in Android 26. It has been around on iOS for quite some time though, back around iOS 7.0.

My goal with this plugin is to allow everyone in flutter to track where the speech is currently at. This can be used to print words on the screen as they are spoken, highlight words of a paragraph as they are uttered, for timing how long each word takes to say at a given speed, or really any other weird reason you may need it. This API is powerful, and I know people are asking for it, and there is simply not a plugin yet that covers it. Surprise! Now there is. You are welcome.

## Features

- [x] Android & iOS
  - [x] speak
  - [x] stop
  - [x] get languages
  - [x] set language
  - [x] is language available
  - [x] set speech rate
  - [x] set speech volume
  - [x] set speech pitch
  - [x] get voices
  - [x] set voice
  - [x] #NEW track what word is currently being said (on all iOS 7+ and Android 26+)
- [x] Android
  - [x] set Silence

## Usage

## Android

Change the minimum Android sdk version to 21 (or higher) in your `android/app/build.gradle` file.

```java
minSdkVersion 21
```

## iOS

There's a known issue with integrating plugins that use Swift into a Flutter project created with the Objective-C template. [Flutter#16049](https://github.com/flutter/flutter/issues/16049)

[Example](https://github.com/loushou/flutter_tts_improved/blob/master/example/lib/main.dart)

To use this plugin :

- add the dependency to your [pubspec.yaml](https://github.com/loushou/flutter_tts_improved/blob/master/example/pubspec.yaml) file.

```yaml
  dependencies:
    flutter:
      sdk: flutter
    flutter_tts_improved:
```

- instantiate a FlutterTtsImproved instance

```dart
FlutterTtsImproved tts = new FlutterTtsImproved();
```

### speak, stop, getLanguages, setLanguage, setSpeechRate, setVolume, setPitch, isLanguageAvailable

```dart
Future _speak() async{
    var result = await tts.speak("Hello World");
    if (result == 1) setState(() => ttsState = TtsState.playing);
}

Future _stop() async{
    var result = await tts.stop();
    if (result == 1) setState(() => ttsState = TtsState.stopped);
}

List<dynamic> languages = await tts.getLanguages;

await tts.setLanguage("en-US");

await tts.setSpeechRate(1.0);

await tts.setVolume(1.0);

await tts.setPitch(1.0);

await tts.isLanguageAvailable("en-US");
```

### Listening for platform calls

```dart
// #NEW
tts.setProgressHandler((String fullPhrase, int wordOffset, int endOfWordOffset, String currentWord) {
  setState(() {
    _wordToDisplay = currentWord;
  });
});

tts.setStartHandler(() {
  setState(() {
    ttsState = TtsState.playing;
  });
});

tts.setCompletionHandler(() {
  setState(() {
    ttsState = TtsState.stopped;
  });
});

tts.setErrorHandler((msg) {
  setState(() {
    ttsState = TtsState.stopped;
  });
});
```

## Getting Started

For help getting started with Flutter, view our online
[documentation](https://flutter.io/).

For help on editing plugin code, view the [documentation](https://flutter.io/platform-plugins/#edit-code).


CHANGELOG.md
## 1.0.3

* Formatted code according to dart pub specs

## 1.0.2

* Fully tested on multiple iOS and Android devices and emulators

## 0.0.2

* Fixed issue with iOS config that prevented plugin from working on iOS devices

## 0.0.1

* Added the usage of the built-in Android and iOS Utterance APIs, when available
* Added stub for older android APIs when Utterance API is not available (pre-android-26)


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_tts_improved_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_tts_improved_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_tts_improved/flutter_tts_improved.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_tts_improved');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('setSpeechRate', () async {
    FlutterTtsImproved tts = FlutterTtsImproved();
    expect(await tts.setSpeechRate(1), '1');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_tts_improved/flutter_tts_improved.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';
  FlutterTtsImproved tts = FlutterTtsImproved();

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    print('VOICES: ${await tts.getVoices}');
    print('LANGUAGES: ${await tts.getLanguages}');

    tts.setProgressHandler((String words, int start, int end, String word) {
      setState(() {
        _platformVersion = word;
      });
      print('PROGRESS: $word => $start - $end');
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              Padding(
                padding: const EdgeInsets.only(bottom: 20),
                child: RaisedButton(
                  onPressed: () {
                    tts.speak('say something that we can debug please');
                  },
                  child: Text('Say Something...'),
                ),
              ),
              Text('Running on: $_platformVersion\n'),
            ],
          ),
        ),
      ),
    );
  }
}


lib/flutter_tts_improved.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

typedef ErrorHandler = void Function(dynamic message);
typedef FlutterTTSImprovedProgressHandler = void Function(
    String, int, int, String);

// Provides Platform specific TTS services (Android: TextToSpeech, IOS: AVSpeechSynthesizer)
class FlutterTtsImproved {
  FlutterTtsImproved() {
    _channel.setMethodCallHandler(platformCallHandler);
  }

  static const MethodChannel _channel = MethodChannel('flutter_tts_improved');

  VoidCallback initHandler;
  VoidCallback startHandler;
  VoidCallback completionHandler;
  FlutterTTSImprovedProgressHandler progressHandler;
  ErrorHandler errorHandler;

  /// [Future] which invokes the platform specific method for speaking
  Future<dynamic> speak(String text) =>
      _channel.invokeMethod<dynamic>('speak', text);

  /// [Future] which invokes the platform specific method for setLanguage
  Future<dynamic> setLanguage(String language) =>
      _channel.invokeMethod<dynamic>('setLanguage', language);

  /// [Future] which invokes the platform specific method for setSpeechRate
  /// Allowed values are in the range from 0.0 (silent) to 1.0 (loudest)
  Future<dynamic> setSpeechRate(double rate) =>
      _channel.invokeMethod<dynamic>('setSpeechRate', rate);

  /// [Future] which invokes the platform specific method for setVolume
  /// Allowed values are in the range from 0.0 (silent) to 1.0 (loudest)
  Future<dynamic> setVolume(double volume) =>
      _channel.invokeMethod<dynamic>('setVolume', volume);

  /// [Future] which invokes the platform specific method for setPitch
  /// 1.0 is default and ranges from .5 to 2.0
  Future<dynamic> setPitch(double pitch) =>
      _channel.invokeMethod<dynamic>('setPitch', pitch);

  /// [Future] which invokes the platform specific method for setVoice
  /// ***Android supported only***
  Future<dynamic> setVoice(String voice) =>
      _channel.invokeMethod<dynamic>('setVoice', voice);

  /// [Future] which invokes the platform specific method for stop
  Future<dynamic> stop() => _channel.invokeMethod<dynamic>('stop');

  /// [Future] which invokes the platform specific method for getLanguages
  /// Android issues with API 21 & 22
  /// Returns a list of available languages
  Future<List<String>> get getLanguages async {
    final List<String> languages =
        List<String>.from(await _channel.invokeMethod<dynamic>('getLanguages'));
    return languages;
  }

  /// [Future] which invokes the platform specific method for getVoices
  /// ***Android supported only ***
  /// Returns a `List` of voice names
  Future<dynamic> get getVoices async {
    final List<String> voices =
        List<String>.from(await _channel.invokeMethod<dynamic>('getVoices'));
    return voices;
  }

  /// [Future] which invokes the platform specific method for isLanguageAvailable
  /// Returns `true` or `false`
  Future<dynamic> isLanguageAvailable(String language) =>
      _channel.invokeMethod<dynamic>(
          'isLanguageAvailable', <String, Object>{'language': language});

  /// [Future] which invokes the platform specific method for setSilence
  /// 0 means start the utterance immediately. If the value is greater than zero a silence period in milliseconds is set according to the parameter
  Future<dynamic> setSilence(int timems) =>
      _channel.invokeMethod<dynamic>('setSilence', timems ?? 0);

  void setStartHandler(VoidCallback callback) {
    startHandler = callback;
  }

  void setCompletionHandler(VoidCallback callback) {
    completionHandler = callback;
  }

  void setProgressHandler(FlutterTTSImprovedProgressHandler callback) {
    progressHandler = callback;
  }

  void setErrorHandler(ErrorHandler handler) {
    errorHandler = handler;
  }

  void ttsInitHandler(VoidCallback handler) {
    initHandler = handler;
  }

  /// Platform listeners
  Future<void> platformCallHandler(MethodCall call) async {
    switch (call.method) {
      case 'tts.init':
        if (initHandler != null) {
          initHandler();
        }
        break;
      case 'speak.onStart':
        if (startHandler != null) {
          startHandler();
        }
        break;
      case 'speak.onComplete':
        if (completionHandler != null) {
          completionHandler();
        }
        break;
      case 'speak.onProgress':
        if (progressHandler != null) {
          final Map<dynamic, dynamic> args = call.arguments;
          progressHandler(
            args['string'].toString(),
            int.parse(args['start']),
            int.parse(args['end']),
            args['word'].toString(),
          );
        }
        break;
      case 'speak.onError':
        if (errorHandler != null) {
          errorHandler(call.arguments);
        }
        break;
      default:
        print('Unknowm method ${call.method}');
    }
  }
}


ios/Classes/SwiftFlutterTtsImprovedPlugin.swift
import Flutter
import UIKit
import AVFoundation

public class SwiftFlutterTtsImprovedPlugin: NSObject, FlutterPlugin, AVSpeechSynthesizerDelegate {
  let synthesizer = AVSpeechSynthesizer()
  var language: String = AVSpeechSynthesisVoice.currentLanguageCode()
  var rate: Float = AVSpeechUtteranceDefaultSpeechRate
  var languages = Set<String>()
  var volume: Float = 1.0
  var pitch: Float = 1.0
  var voice: AVSpeechSynthesisVoice?

  var channel = FlutterMethodChannel()

  init(channel: FlutterMethodChannel) {
    super.init()
    self.channel = channel
    synthesizer.delegate = self
    setLanguages()
    
    // Allow audio playback when the Ring/Silent switch is set to silent
    do {
      try AVAudioSession.sharedInstance().setCategory(.playback)
      try AVAudioSession.sharedInstance().setActive(true)
    } catch {
      print(error)
    }
  }

  private func setLanguages() {
    for voice in AVSpeechSynthesisVoice.speechVoices(){
      self.languages.insert(voice.language)
    }
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_tts_improved", binaryMessenger: registrar.messenger())
    let instance = SwiftFlutterTtsImprovedPlugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "speak":
      let text: String = call.arguments as! String
      self.speak(text: text)
      result(1)
      break
    case "setLanguage":
      let language: String = call.arguments as! String
      self.setLanguage(language: language, result: result)
      break
    case "setSpeechRate":
      let rate: Double = call.arguments as! Double
      self.setRate(rate: Float(rate))
      result(1)
      break
    case "setVolume":
      let volume: Double = call.arguments as! Double
      self.setVolume(volume: Float(volume), result: result)
      break
    case "setPitch":
      let pitch: Double = call.arguments as! Double
      self.setPitch(pitch: Float(pitch), result: result)
      break
    case "stop":
      self.stop()
      result(1)
      break
    case "getLanguages":
      self.getLanguages(result: result)
      break
    case "isLanguageAvailable":
      let arg: Dictionary<String, String> = call.arguments as! Dictionary<String, String>
      self.isLanguageAvailable(language: arg["language"]! as String, result: result)
      break
    case "getVoices":
      self.getVoices(result: result)
      break
    case "setVoice":
      let voiceName = call.arguments as! String
      self.setVoice(voiceName: voiceName, result: result)
      break
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func speak(text: String) {
    let utterance = AVSpeechUtterance(string: text)
    if self.voice != nil {
      utterance.voice = self.voice!
    } else {
      utterance.voice = AVSpeechSynthesisVoice(language: self.language)
    }
    utterance.rate = self.rate
    utterance.volume = self.volume
    utterance.pitchMultiplier = self.pitch
    
    self.synthesizer.speak(utterance)
  }

  private func setLanguage(language: String, result: FlutterResult) {
    if !(self.languages.contains(language)) {
      result(0)
    } else {
      self.language = language
      self.voice = nil
      result(1)
    }
  }

  private func setRate(rate: Float) {
    self.rate = rate
  }

  private func setVolume(volume: Float, result: FlutterResult) {
    if (volume >= 0.0 && volume <= 1.0) {
      self.volume = volume
      result(1)
    } else {
      result(0)
    }
  }

  private func setPitch(pitch: Float, result: FlutterResult) {
    if (volume >= 0.5 && volume <= 2.0) {
      self.pitch = pitch
      result(1)
    } else {
      result(0)
    }
  }

  private func stop() {
    self.synthesizer.stopSpeaking(at: AVSpeechBoundary.immediate)
  }

  private func getLanguages(result: FlutterResult) {
    result(Array(self.languages))
  }

  private func isLanguageAvailable(language: String, result: FlutterResult) {
    var isAvailable: Bool = false
    if (self.languages.contains(language)) {
      isAvailable = true
    }
    result(isAvailable);
  }

  private func getVoices(result: FlutterResult) {
    if #available(iOS 9.0, *) {
      let voices = NSMutableArray()
      for voice in AVSpeechSynthesisVoice.speechVoices() {
        voices.add(voice.name)
      }
      result(voices)
    } else {
      // Since voice selection is not supported below iOS 9, make voice getter and setter
      // have the same bahavior as language selection.
      getLanguages(result: result)
    }
  }

  private func setVoice(voiceName: String, result: FlutterResult) {
    if #available(iOS 9.0, *) {
      if let voice = AVSpeechSynthesisVoice.speechVoices().first(where: { $0.name == voiceName }) {
        self.voice = voice
        self.language = voice.language
        result(1)
        return
      }
      result(0)
    } else {
      setLanguage(language: voiceName, result: result)
    }
  }

  public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance) {
    self.channel.invokeMethod("speak.onComplete", arguments: nil)
  }

  public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didStart utterance: AVSpeechUtterance) {
    self.channel.invokeMethod("speak.onStart", arguments: nil)
  }

  public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, willSpeakRangeOfSpeechString characterRange: NSRange, utterance: AVSpeechUtterance) {
    let nsWord = utterance.speechString as NSString
    let data: [String:String] = [
      "string": utterance.speechString,
      "start": String(characterRange.location),
      "end": String(characterRange.location + characterRange.length),
      "word": nsWord.substring(with: characterRange)
    ]
    self.channel.invokeMethod("speak.onProgress", arguments: data)
  }
}

// Helper function inserted by Swift 4.2 migrator.
fileprivate func convertFromAVAudioSessionCategory(_ input: AVAudioSession.Category) -> String {
	return input.rawValue
}


android/src/main/java/com/rdygo/flutter_tts_improved/FlutterTtsImprovedPlugin.java
package com.rdygo.flutter_tts_improved;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.speech.tts.Voice;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Locale;
import java.util.UUID;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * FlutterTtsPlugin
 */
public class FlutterTtsImprovedPlugin implements MethodCallHandler {
  private final Handler handler;
  private final MethodChannel channel;
  private TextToSpeech tts;
  private final String tag = "TTS";
  private final String googleTtsEngine = "com.google.android.tts";
  String uuid;
  Bundle bundle;
  private int silencems;
  private static final String SILENCE_PREFIX = "SIL_";
  private final HashMap<String, String> utterances = new HashMap<>();

  /**
   * Plugin registration.
   */
  private FlutterTtsImprovedPlugin(Context context, MethodChannel channel) {
    this.channel = channel;
    this.channel.setMethodCallHandler(this);

    handler = new Handler(Looper.getMainLooper());
    bundle = new Bundle();
    tts = new TextToSpeech(context.getApplicationContext(), onInitListener, googleTtsEngine);
  }

  private UtteranceProgressListener utteranceProgressListener =
    new UtteranceProgressListener() {
      @Override
      public void onStart(String utteranceId) {
        invokeMethod("speak.onStart", true);
        if (Build.VERSION.SDK_INT < 26) {
          this.triggerProgress(utteranceId, 0, utterances.get(utteranceId).length());
        }
      }

      @Override
      public void onDone(String utteranceId) {
        if (utteranceId != null && utteranceId.startsWith(SILENCE_PREFIX))
          return;
        invokeMethod("speak.onComplete", true);
        utterances.remove(utteranceId);
      }

      private void triggerProgress(String utteranceId, int startAt, int endAt) {
        final String words = utterances.get(utteranceId);
        final HashMap<String, String> data = new HashMap<>();
        data.put("string", words);
        data.put("start", Integer.toString(startAt));
        data.put("end", Integer.toString(endAt));
        data.put("word", words.substring(startAt, endAt));
        invokeMethod("speak.onProgress", data);
      }

      // only valid on android 26 and higher
      @Override
      public void onRangeStart(String utteranceId, int startAt, int endAt, int frame) {
        super.onRangeStart(utteranceId, startAt, endAt, frame);
        this.triggerProgress(utteranceId, startAt, endAt);
      }

      @Override
      @Deprecated
      public void onError(String utteranceId) {
        invokeMethod("speak.onError", "Error from TextToSpeech");
      }

      @Override
      public void onError(String utteranceId, int errorCode) {
        invokeMethod("speak.onError", "Error from TextToSpeech - " + errorCode);
      }
    };

  private TextToSpeech.OnInitListener onInitListener =
    new TextToSpeech.OnInitListener() {
      @Override
      public void onInit(int status) {
        if (status == TextToSpeech.SUCCESS) {
          tts.setOnUtteranceProgressListener(utteranceProgressListener);
          invokeMethod("tts.init", true);

          try {
            Locale locale = tts.getDefaultVoice().getLocale();
            if (isLanguageAvailable(locale)) {
              tts.setLanguage(locale);
            }
          } catch (NullPointerException | IllegalArgumentException e) {
            Log.e(tag, "getDefaultLocale: " + e.getMessage());
          }
        } else {
          Log.e(tag, "Failed to initialize TextToSpeech");
        }
      }
    };

  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_tts_improved");
    channel.setMethodCallHandler(new FlutterTtsImprovedPlugin(registrar.activeContext(), channel));
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("speak")) {
      String text = call.arguments.toString();
      speak(text);
      result.success(1);
    } else if (call.method.equals("stop")) {
      stop();
      result.success(1);
    } else if (call.method.equals("setSpeechRate")) {
      String rate = call.arguments.toString();
      setSpeechRate(Float.parseFloat(rate));
      result.success(1);
    } else if (call.method.equals("setVolume")) {
      String volume = call.arguments.toString();
      setVolume(Float.parseFloat(volume), result);
    } else if (call.method.equals("setPitch")) {
      String pitch = call.arguments.toString();
      setPitch(Float.parseFloat(pitch), result);
    } else if (call.method.equals("setLanguage")) {
      String language = call.arguments.toString();
      setLanguage(language, result);
    } else if (call.method.equals("getLanguages")) {
      getLanguages(result);
    } else if (call.method.equals("getVoices")) {
      getVoices(result);
    } else if (call.method.equals("setVoice")) {
      String voice = call.arguments.toString();
      setVoice(voice, result);
    } else if (call.method.equals("isLanguageAvailable")) {
      String language = ((HashMap) call.arguments()).get("language").toString();
      Locale locale = Locale.forLanguageTag(language);
      result.success(isLanguageAvailable(locale));
    } else if (call.method.equals("setSilence")) {
      String silencems = call.arguments.toString();
      this.silencems = Integer.parseInt(silencems);
    } else {
      result.notImplemented();
    }
  }

  void setSpeechRate(float rate) {
    tts.setSpeechRate(rate);
  }

  Boolean isLanguageAvailable(Locale locale) {
    return tts.isLanguageAvailable(locale) >= TextToSpeech.LANG_AVAILABLE;
  }

  void setLanguage(String language, Result result) {
    Locale locale = Locale.forLanguageTag(language);
    if (isLanguageAvailable(locale)) {
      tts.setLanguage(locale);
      result.success(1);
    } else {
      result.success(0);
    }
  }

  void setVoice(String voice, Result result) {
    for (Voice ttsVoice : tts.getVoices()) {
      if (ttsVoice.getName().equals(voice)) {
        tts.setVoice(ttsVoice);
        result.success(1);
        return;
      }
    }
    Log.d(tag, "Voice name not found: " + voice);
    result.success(0);
  }

  void setVolume(float volume, Result result) {
    if (volume >= 0.0F && volume <= 1.0F) {
      bundle.putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, volume);
      result.success(1);
    } else {
      Log.d(tag, "Invalid volume " + volume + " value - Range is from 0.0 to 1.0");
      result.success(0);
    }
  }

  void setPitch(float pitch, Result result) {
    if (pitch >= 0.5F && pitch <= 2.0F) {
      tts.setPitch(pitch);
      result.success(1);
    } else {
      Log.d(tag, "Invalid pitch " + pitch + " value - Range is from 0.5 to 2.0");
      result.success(0);
    }
  }

  void getVoices(Result result) {
    ArrayList<String> voices = new ArrayList<>();
    try {
      for (Voice voice : tts.getVoices()) {
        voices.add(voice.getName());
      }
      result.success(voices);
    } catch (NullPointerException e) {
      Log.d(tag, "getVoices: " + e.getMessage());
      result.success(null);
    }
  }

  void getLanguages(Result result) {
    ArrayList<String> locales = new ArrayList<>();
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      // While this method was introduced in API level 21, it seems that it
      // has not been implemented in the speech service side until API Level 23.
      for (Locale locale : tts.getAvailableLanguages()) {
        locales.add(locale.toLanguageTag());
      }
    } else {
      for (Locale locale : Locale.getAvailableLocales()) {
        if (locale.getVariant().isEmpty() && isLanguageAvailable(locale)) {
          locales.add(locale.toLanguageTag());
        }
      }
    }
    result.success(locales);
  }

  private void speak(String text) {
    uuid = UUID.randomUUID().toString();
    utterances.put(uuid, text);
    if (silencems > 0) {
      tts.playSilentUtterance(silencems, TextToSpeech.QUEUE_FLUSH, SILENCE_PREFIX + uuid);
      tts.speak(text, TextToSpeech.QUEUE_ADD, bundle, uuid);
    } else {
      tts.speak(text, TextToSpeech.QUEUE_FLUSH, bundle, uuid);
    }
  }

  private void stop() {
    tts.stop();
  }

  private void invokeMethod(final String method, final Object arguments) {
    handler.post(new Runnable() {
      @Override
      public void run() {
        channel.invokeMethod(method, arguments);
      }
    });
  }
}
